06_puppy/nicholascross/main.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+type ID uint64
+type Breed string
+type Colour string
+type Value float64
+
+type Puppy struct {
+	ID     ID
+	Breed  Breed
+	Colour Colour
+	Value  Value
+}
+
+type Storer interface {
+	CreatePuppy(p Puppy)
+	RetrievePuppy(ID) *Puppy
+	UpdatePuppy(ID, Puppy)
+	DeletePuppy(ID)
+}
+
+type MapStore map[ID]*Puppy
+
+var out io.Writer = os.Stdout
+
+func main() {
+	p := Puppy{1, "Jack Russel Terrier", "white and brown", 550.0}
+	fmt.Fprintf(out, "%d - %s [%s]: %f", p.ID, p.Breed, p.Colour, p.Value)
+}
+
+func (ms MapStore) CreatePuppy(p Puppy) {
+	if _, ok := ms[p.ID]; !ok {
+		ms[p.ID] = &p
+	}
+}
+
+func (ms MapStore) RetrievePuppy(id ID) *Puppy {
```

Comment:
By returning a pointer, you allow the caller to modify the underlying data. You can get around this by copying the stored data into a new variable and returning a pointer to that
```
pup := *ms[id] // dereference and copy raw into new variable
return &pup // return pointer to new variable
```
or using a raw store.